packages/langium/src/lsp/signature-help-provider.ts

```diff
@@ -0,0 +1,78 @@
+
+import { CancellationToken, SignatureHelp, SignatureHelpOptions, SignatureHelpParams } from 'vscode-languageserver';
+import { AstNode } from '../syntax-tree';
+import { findLeafNodeAtOffset, findRelevantNode } from '../utils/cst-util';
+import { MaybePromise } from '../utils/promise-util';
+import { LangiumDocument } from '../workspace/documents';
+
+/**
+ * Language-specific service for handling signature help requests.
+ */
+export interface SignatureHelpProvider {
+    /**
+     * Handles a signature help request
+     */
+    provideSignatureHelp(document: LangiumDocument, params: SignatureHelpParams, cancelToken?: CancellationToken): MaybePromise<SignatureHelp | undefined>;
+    /**
+     * Options that determine the server capabilities for a signature help request. It contains the list of triggering characters.
+     */
+    get signatureHelpOptions(): SignatureHelpOptions;
+}
+
+export abstract class AbstractSignatureHelpProvider implements SignatureHelpProvider {
+    provideSignatureHelp(document: LangiumDocument, params: SignatureHelpParams): MaybePromise<SignatureHelp | undefined> {
+        const rootNode = document.parseResult.value;
+        const cst = rootNode.$cstNode;
+        if (cst) {
+            const sourceCstNode = findLeafNodeAtOffset(cst, document.textDocument.offsetAt(params.position));
+            if (sourceCstNode) {
+                const element = findRelevantNode(sourceCstNode);
+                if (element) {
+                    return this.getSignatureFromElement(element);
+                }
+            }
+        }
+        return undefined;
+    }
+
+    /**
+     * Override this method to return the desired SignatureHelp
+     */
+    protected abstract getSignatureFromElement(element: AstNode): MaybePromise<SignatureHelp | undefined>;
+
+    /**
+     * Override this getter to return the list of triggering characters for your language. To deactivate the signature help, return an empty object.
+     */
+    get signatureHelpOptions(): SignatureHelpOptions {
+        return { triggerCharacters: ['('] };
+    }
+}
+
+/**
+ * Merges the SignatureHelpOptions of all languages
+ */
+export function mergeSignatureHelpOptions(options: Array<SignatureHelpOptions | undefined>): SignatureHelpOptions | undefined {
+    const triggerCharacters: string[] = [];
+    const retriggerCharacters: string[] = [];
+
+    options.forEach(option => {
+        if (option?.triggerCharacters) {
+            triggerCharacters.push(...option.triggerCharacters);
+        }
+        if (option?.retriggerCharacters) {
+            retriggerCharacters.push(...option.retriggerCharacters);
+        }
+    });
+
+    const mergedOptions: SignatureHelpOptions = {
+        triggerCharacters: triggerCharacters.length > 0 ? Array.from(new Set<string>(triggerCharacters)) : undefined,
+        retriggerCharacters: retriggerCharacters.length > 0 ? Array.from(new Set<string>(retriggerCharacters)) : undefined
```

Comment:
Minor: TypeScript automatically infers the type of the set. We should also sort the arrays to better test them
```suggestion
        triggerCharacters: triggerCharacters.length > 0 ? Array.from(new Set(triggerCharacters)).sort() : undefined,
        retriggerCharacters: retriggerCharacters.length > 0 ? Array.from(new Set(retriggerCharacters)).sort() : undefined
```